File: backend/src/models/Portfolio.ts
import mongoose, { Schema, Document } from "mongoose";

export interface IPortfolio extends Document {
  userId: string;
  holdings: Array<{
    symbol: string;
    quantity: number;
    averagePrice: number;
  }>;
  cash: number;
}

const PortfolioSchema: Schema = new Schema({
  userId: { type: String, required: true, unique: true },
  holdings: [
    {
      symbol: { type: String, required: true },
      quantity: { type: Number, required: true },
      averagePrice: { type: Number, required: true },
    },
  ],
  cash: { type: Number, default: 10000 },
});

export default mongoose.model<IPortfolio>("Portfolio", PortfolioSchema);

File: backend/src/engine/models/portfolio.py
class Portfolio:
    def __init__(self, user_id):
        self.user_id = user_id
        self.holdings = []
        self.cash = 10000.0

    def add_holding(self, symbol, quantity, price):
        for holding in self.holdings:
            if holding['symbol'] == symbol:
                total_quantity = holding['quantity'] + quantity
                total_cost = (holding['quantity'] * holding['averagePrice']) + (quantity * price)
                holding['averagePrice'] = total_cost / total_quantity
                holding['quantity'] = total_quantity
                return
        self.holdings.append({
            'symbol': symbol,
            'quantity': quantity,
            'averagePrice': price
        })

    def to_dict(self):
        return {
            'user_id': self.user_id,
            'holdings': self.holdings,
            'cash': self.cash
        }